### PR TITLE
0D, 1D and zero volume support for normalization ops

### PR DESCRIPTION
**Pull request by nsorabaTT**
### Ticket
[Addresses Issue 18362](https://github.com/tenstorrent/tt-metal/issues/18362)

### Problem description
Currently, normalization operations only support non-zero volume tensors and tensors with rank >=2.

### What's changed
This PR adds the following:

1. Support 0D and 1D tensors for softmax, layer_norm and rms_norm
2. Support zero volume tensors of any rank (0-4).
3. Support 0V tensors for fused softmax ops.
4. Add sweeps to compare ttnn behavior w.r.t torch for all the flavors above.

| Reduce Op | All ranks supported  | Zero volume supported? | Matches Torch?  | 
|-----------|----------------------|------------------------|-----------------|
| `softmax`     |          Y           |            Y           |        Y        |
| `rms_norm`    |          Y           |            Y           |        Y        |
| `layer_norm`     |          Y           |            Y           |        Y        |
| `softmax in-place`     |          Y           |            N           |        Y        |
| `softmax fused`     |          N           |            Y           |        Y        |
| `batch_norm`*     |          N           |            Y           |        Y        |
| `group_norm`*     |          N           |            Y           |        Y        |

*Batch Norm and Group Norm apply for tensors with rank >= 2. Thus, 0D/1D support does not arise.

### Checklist
- [x] [All post commit #28933](https://github.com/tenstorrent/tt-metal/actions/runs/14207527446) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [x] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [x] New/Existing tests provide coverage for changes

#### Runs of new sweeps
[2471](https://github.com/tenstorrent/tt-metal/actions/runs/14207279598)
[2472](https://github.com/tenstorrent/tt-metal/actions/runs/14207279598)
[2473](https://github.com/tenstorrent/tt-metal/actions/runs/14207283992)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/normalization/rmsnorm/rmsnorm.cpp [hunk 1]
```diff
@@ -17,6 +21,30 @@ ttnn::Tensor ExecuteRMSNorm::invoke(
...
+
+    // For 0V tensors
+    if (input_tensor.get_logical_volume() == 0) [[unlikely]] {
+        return ttnn::clone(input_tensor, /*dtype=*/std::nullopt, output_memory_config, compute_kernel_config);
+    }
+
+    // For 0D tensors
+    if (rank == 0) [[unlikely]] {
+        float fill_val = 1.0f;
+        // TODO: fill_val needs to be -1.0f when the 0D value is negative
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/normalization/rmsnorm/rmsnorm.cpp:35]: If this will be done in a future PR, please reference an issue (either a current one or a new one).

File: tests/sweep_framework/sweeps/normalization/generality/softmax_fused.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    ttnn_result = ttnn.to_torch(ttnn.from_device(ttnn_result))
+
+    pcc_result, msg = check_with_pcc(torch_result, ttnn_result, 0.99)
+
+    if not pcc_result:
+        return [False, msg]
+
+    atol = rtol = 0.1
+
+    # print(
```
Comment (bbradelTT) [inline tests/sweep_framework/sweeps/normalization/generality/softmax_fused.py:107]: Should be removed or uncommented (in which case should probably use logger).

File: tests/sweep_framework/sweeps/normalization/generality/softmax_fused.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+    torch_mask = torch_tensor.clone() < 0
+    ttnn_mask = ttnn.from_torch(torch_mask.to(torch.bfloat16), layout=ttnn.TILE_LAYOUT, device=device)
+
+    scale = random.uniform(0.1, 10.0)
+    has_scale_mask = ("scale" in op) and ("mask" in op)
+    torch_op, ttnn_op = torch.softmax, getattr(ttnn, op)
+
+    # Print tensor and scale
+    print(
```
Comment (bbradelTT) [inline tests/sweep_framework/sweeps/normalization/generality/softmax_fused.py:62]: Could you please try logger.info?

**Review by bbradelTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/normalization/rmsnorm/rmsnorm.cpp (hunk 1 above)
Comment (nsorabaTT) [reply ttnn/cpp/ttnn/operations/normalization/rmsnorm/rmsnorm.cpp]: Thank you. Was a stale TODO, removed it.

File: tests/sweep_framework/sweeps/normalization/generality/softmax_fused.py (hunk 3 above)
Comment (nsorabaTT) [reply tests/sweep_framework/sweeps/normalization/generality/softmax_fused.py]: Done

File: tests/sweep_framework/sweeps/normalization/generality/softmax_fused.py (hunk 2 above)
Comment (nsorabaTT) [reply tests/sweep_framework/sweeps/normalization/generality/softmax_fused.py]: No longer needed, removed.

**Review by vsureshTT — APPROVED** [review]

File: tests/sweep_framework/sweeps/normalization/generality/softmax.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        torch_result *= scale
+
+    pcc_result, msg = check_with_pcc(torch_result, ttnn_result, 0.99)
+
+    if not pcc_result:
+        return [False, msg]
+
+    atol = rtol = 0.1
+
+    return [
```
Comment (jdesousa-TT) [inline tests/sweep_framework/sweeps/normalization/generality/softmax.py:95]: Same as above

File: tests/sweep_framework/sweeps/normalization/generality/softmax_fused.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    ttnn_result = ttnn.to_torch(ttnn.from_device(ttnn_result))
+
+    pcc_result, msg = check_with_pcc(torch_result, ttnn_result, 0.99)
+
+    if not pcc_result:
+        return [False, msg]
+
+    atol = rtol = 0.1
+
+    return [
```
Comment (jdesousa-TT) [inline tests/sweep_framework/sweeps/normalization/generality/softmax_fused.py:109]: Same

**Review by jdesousa-TT — CHANGES_REQUESTED** [review]

File: tests/sweep_framework/sweeps/normalization/generality/layernorm.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    ttnn_result = ttnn.to_torch(ttnn.from_device(ttnn_result))
+
+    pcc_result, msg = check_with_pcc(torch_result, ttnn_result, 0.99)
+
+    if not pcc_result:
+        return [False, msg]
+
+    atol = rtol = 0.1
+
+    return [
```
[the remark this answers is not in the document]
Comment (nsorabaTT) [reply tests/sweep_framework/sweeps/normalization/generality/layernorm.py]: FIxed

File: tests/sweep_framework/sweeps/normalization/generality/softmax_fused.py (hunk 5 above)
Comment (nsorabaTT) [reply tests/sweep_framework/sweeps/normalization/generality/softmax_fused.py]: FIxed

**Review by jdesousa-TT — APPROVED** [review]
